server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -233,20 +270,19 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
             assert startingSeqNo >= 0 : "startingSeqNo must be non negative. got: " + startingSeqNo;
 
             sendFileStep.whenComplete(r -> {
-                if (shard.indexSettings().isSoftDeleteEnabled()
-                    && shard.indexSettings().getIndexMetaData().getState() != IndexMetaData.State.CLOSE) {
+                if (useRetentionLeases && isSequenceNumberBasedRecovery == false) {
+                    // We can in general use retention leases for peer recovery, but there is no lease for the target node right now.
                     runUnderPrimaryPermit(() -> {
-                        try {
-                            // conservative estimate of the GCP for creating the lease. TODO use the actual GCP once it's appropriate
-                            final long globalCheckpoint = startingSeqNo - 1;
-                            // blindly create the lease. TODO integrate this with the recovery process
-                            shard.addPeerRecoveryRetentionLease(request.targetNode().getId(), globalCheckpoint,
+                            // Start the lease off retaining all the history needed from the local checkpoint of the safe commit that we've
+                            // just established on the replica. This primary is certainly retaining such history, but other replicas might
+                            // not be. No big deal if this recovery succeeds, but if this primary fails then the new primary might have to
+                            // repeat phase 1 to recover this replica.
+                            // TODO TBD maybe do this earlier?
```

Comment:
Should we do this earlier? We could do it before phase 1, with the advantage that it gives the other replicas a fighting chance to also retain the history needed to resume this recovery on a primary failure (still not guaranteed, mind) but the disadvantage of retaining extra history if this recovery fails and then we recover to a different node.

FWIW I prefer it here.